wgengine/packet/ip.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package packet
+
+import (
+	"fmt"
+	"net"
+)
+
+// IP is an IPv4 address.
+type IP uint32
+
+// NewIP converts a standard library IP address into an IP.
+// It panics if b is not an IPv4 address.
+func NewIP(b net.IP) IP {
+	b4 := b.To4()
+	if b4 == nil {
+		panic(fmt.Sprintf("To4(%v) failed", b))
+	}
+	return IP(get32(b4))
+}
+
+func (ip IP) String() string {
+	return fmt.Sprintf("%d.%d.%d.%d", byte(ip>>24), byte(ip>>16), byte(ip>>8), byte(ip))
+}
+
+// IPProto encodes an IP protocol, such as TCP or UDP.
+type IPProto uint8
+
+// IPProto is either a real IP protocol (ICMP, TCP, UDP, ...) or an flag like Junk.
+// If it is a real IP protocol, its value corresponds to its IP protocol number.
+// TODO(dmytro): flags should be taken out of here.
+const (
+	// Junk represents an unknown or unsupported protocol; it's deliberately the zero value.
+	Junk IPProto = 0x00
+	ICMP IPProto = 0x01
+	TCP  IPProto = 0x06
+	UDP  IPProto = 0x11
+	// 0xFE and 0xFF are unassigned.
+	IPv6     IPProto = 0xFE
+	Fragment IPProto = 0xFF
+)
+
+func (p IPProto) String() string {
+	switch p {
+	case Fragment:
+		return "Frag"
+	case ICMP:
+		return "ICMP"
+	case UDP:
+		return "UDP"
+	case TCP:
+		return "TCP"
+	case IPv6:
+		return "IPv6"
+	default:
+		return "Junk"
+	}
+}
+
+// IPHeader represents an IP packet header.
+type IPHeader struct {
+	IPProto IPProto
+	IPID    uint16
+	SrcIP   IP
+	DstIP   IP
+}
+
+const ipHeaderLength = 20
+
+func (h IPHeader) Length() int {
+	return ipHeaderLength
+}
+
+func (h IPHeader) Marshal(buf []byte) error {
+	if len(buf) < ipHeaderLength {
+		return errSmallBuffer
+	}
+	length := len(buf)
```

Comment:
I was wondering why you pulled this up into a variable instead of just using len(buf) below, but then I realized you probably were about to do a bounds check that it's small enough to fit in a uint16 and return an error? Or cap it at 64K?